scalding-core/src/main/scala/com/twitter/scalding/Stats.scala

```diff
@@ -125,9 +125,15 @@ object RuntimeStats extends java.io.Serializable {
   def addFlowProcess(fp: FlowProcess[_]) {
     val uniqueJobIdObj = fp.getProperty(UniqueID.UNIQUE_JOB_ID)
     if (uniqueJobIdObj != null) {
-      uniqueJobIdObj.asInstanceOf[String].split(",").foreach { uniqueId =>
+      // for speed concern, use a while loop instead of foreach here
+      val splitted = StringUtility.fastSplit(uniqueJobIdObj.asInstanceOf[String], ",")
+      val size = splitted.length
+      var i = 0
+      while (i < size) {
+        val uniqueId = splitted(i)
         logger.debug("Adding flow process id: " + uniqueId)
-        flowMappingStore.put(uniqueId, new WeakReference(fp))
+        flowMappingStore.getOrElse(uniqueId, new WeakReference(fp))
```

Comment:
My fault: I was wrong. .contains won't work here because the WeakReference may be out of scope (and is in the tests), I think. We just need to put each time.

To save the work, maybe keep a
```scala
private[this] var prevFP: FlowProcess = null
```
and `if (prevFP == fp)` don't do any work. This will be the common case (especially in the WrappedJoiner), so the branch predictor should see that is almost always true, and it should be almost free for users that don't use stats.